examples/pytorch/deepspeed-demo/train_bert_ds.py

```diff
@@ -0,0 +1,829 @@
+"""
+Modified version of train_bert.py that adds DeepSpeed
+"""
+
+import datetime
+import json
+import logging
+import os
+import pathlib
+import random
+import re
+import string
+from functools import partial
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, TypeVar, Union
+
+import datasets
+import deepspeed
+import fire
+import loguru
+import numpy as np
+import pytz
+import sh
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader, Dataset
+from torch.utils.tensorboard import SummaryWriter
+from transformers import AutoTokenizer, PreTrainedTokenizer, PreTrainedTokenizerFast
+from transformers.models.roberta import RobertaConfig, RobertaModel
+from transformers.models.roberta.modeling_roberta import (
+    RobertaLMHead,
+    RobertaPreTrainedModel,
+)
+
+
+def is_rank_0() -> bool:
+    return int(os.environ.get("RANK", "0")) == 0
+
+
+# Logging Functions
+
+logger = loguru.logger
+
+
+def log_dist(message: str, ranks: List[int] = [], level: int = logging.INFO) -> None:
+    """Log messages for specified ranks only"""
+    my_rank = int(os.environ.get("RANK", "0"))
+    if my_rank in ranks:
+        if level == logging.INFO:
+            logger.info(f"[Rank {my_rank}] {message}")
+        if level == logging.ERROR:
+            logger.error(f"[Rank {my_rank}] {message}")
+        if level == logging.DEBUG:
+            logger.debug(f"[Rank {my_rank}] {message}")
```

Comment:
If we don't have this check every replica will print logs from all Ranks, right ?

Reply:
Yes, currently only rank 0 (master) will print these logs, for example:
```
[2024-09-06 08:31:09,542] [INFO] [timer.py:198:stop] 0/9980, RunningAvgSamplesPerSec=287.4605218033503, CurrSamplesPerSec=263.40649911882343, MemAllocated=0.06GB, MaxMemAllocated=1.68GB
2024-09-06 08:31:09.543 | INFO     | __main__:log_dist:53 - [Rank 0] Loss: 7.0046
```